apps/passport-server/src/database/queries/pipelineOfflineCheckinDB.ts

```diff
@@ -0,0 +1,119 @@
+import { PipelineOfflineCheckin } from "@pcd/passport-interface";
+import { Pool } from "postgres-pool";
+import { sqlQuery } from "../sqlQuery";
+
+/**
+ * Interface for the storage of offline checkins for asynchronous processing.
+ */
+export interface IPipelineOfflineCheckinDB {
+  addOfflineCheckins(
+    pipelineId: string,
+    checkerEmail: string,
+    ticketId: string[],
+    checkinTimestamp: Date
+  ): Promise<void>;
+  getOfflineCheckinsForPipeline(
+    pipelineId: string
+  ): Promise<PipelineOfflineCheckin[]>;
+  deleteOfflineCheckins(pipelineId: string, ticketIds: string[]): Promise<void>;
+  addFailedOfflineCheckin(pipelineId: string, ticketId: string): Promise<void>;
+}
+
+/**
+ * Manages offline checkins for asynchronous processing.
+ */
+export class PipelineOfflineCheckinDB implements IPipelineOfflineCheckinDB {
+  private db: Pool;
+
+  public constructor(db: Pool) {
+    this.db = db;
+  }
+
+  /**
+   * Add multiple offline check-ins for a given event on a pipeline.
+   */
+  public async addOfflineCheckins(
+    pipelineId: string,
+    checkerEmail: string,
+    ticketIds: string[],
+    checkinTimestamp: Date
+  ): Promise<void> {
+    await sqlQuery(
+      this.db,
+      `INSERT INTO generic_issuance_offline_checkins (pipeline_id, checker_email, ticket_id, checkin_timestamp)
+      VALUES ${ticketIds
+        .map(
+          (_, i) =>
+            `($${i * 4 + 1},$${i * 4 + 2}, $${i * 4 + 3}, $${i * 4 + 4})`
```

Comment:
I don't really like this, because it's very easy to get it wrong, but it seems like most JavaScript SQL templating libraries are either unmaintained (though maybe this is because they're finished and have no bugs) or require you to adopt a whole stack of libraries, including a connection manager and so on.